Destroy vgt instance on exit

Add an exit notifier so that if the guest is forcibly shutdown or
rebooted, it still cleans up the vgt instance that was created (if any).

diff --git a/hw/display/vgt_vga.c b/hw/display/vgt_vga.c
index e570ec1..e92aaeb 100644
--- a/hw/display/vgt_vga.c
+++ b/hw/display/vgt_vga.c
@@ -91,6 +91,10 @@ typedef struct VGTVGAState {
     int domid;
 } VGTVGAState;
 
+static Notifier exit_notifier;
+
+static void vgt_exit_notify(Notifier *n, void *data);
+
 /* These are the default values */
 int vgt_low_gm_sz = 64; /* in MB */
 int vgt_high_gm_sz = 448; /* in MB */
@@ -441,6 +445,8 @@ static void create_vgt_instance(VGTVGAState *vdev)
     }
 
     vdev->instance_created = TRUE;
+    exit_notifier.notify = vgt_exit_notify;
+    qemu_add_exit_notifier(&exit_notifier);
 
     config_vgt_guest_monitors();
 }
@@ -597,6 +603,17 @@ static void vgt_cleanupfn(PCIDevice *dev)
     }
 }
 
+static void vgt_exit_notify(Notifier *n, void *data)
+{
+    /*
+     * The vgt instance may have already been destroyed by the normal
+     * cleanup function. This covers the case where QEMU is forcibly
+     * terminated.
+     */
+    if (vgt_vga_enabled)
+        destroy_vgt_instance(xen_domid);
+}
+
 static int vgt_get_domid(void)
 {
     int domid = 0;
